po-docs/ja/virt-win-reg.1
.\" Automatically generated by Podwrapper::Man 1.40.2 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "virt-win-reg 1"
.TH virt-win-reg 1 "2019-02-07" "libguestfs-1.40.2" "Virtualization Support"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "名前"
.IX Header "名前"
virt-win-reg \- Windows 仮想マシンからの Windows レジストリエントリーのエクスポートおよびマージ
.SH "書式"
.IX Header "書式"
.Vb 1
\& virt\-win\-reg domname \*(AqHKLM\ePath\eTo\eSubkey\*(Aq
\&
\& virt\-win\-reg domname \*(AqHKLM\ePath\eTo\eSubkey\*(Aq name
\&
\& virt\-win\-reg domname \*(AqHKLM\ePath\eTo\eSubkey\*(Aq @
\&
\& virt\-win\-reg \-\-merge domname [input.reg ...]
\&
\& virt\-win\-reg [\-\-options] disk.img ... # 仮想マシン名の代わり
.Ve
.SH "警告"
.IX Header "警告"
You must \fInot\fR use \f(CW\*(C`virt\-win\-reg\*(C'\fR with the \fI\-\-merge\fR option on live
virtual machines.  If you do this, you \fIwill\fR get irreversible disk
corruption in the \s-1VM.\s0  \f(CW\*(C`virt\-win\-reg\*(C'\fR tries to stop you from doing this,
but doesn't catch all cases.
.PP
Modifying the Windows Registry is an inherently risky operation.  The format
is deliberately obscure and undocumented, and Registry changes can leave the
system unbootable.  Therefore when using the \fI\-\-merge\fR option, make sure
you have a reliable backup first.
.SH "説明"
.IX Header "説明"
このプログラムは Windows 仮想マシンから Windows レジストリのエントリーをエクスポートおよびマージできます。
.PP
第一パラメーターは Windows 仮想マシンの libvirt 仮想マシン名または生のディスクイメージです。
.PP
If \fI\-\-merge\fR is \fInot\fR specified, then the chosen registry key is
displayed/exported (recursively).  For example:
.PP
.Vb 1
\& $ virt\-win\-reg Windows7 \*(AqHKEY_LOCAL_MACHINE\eSOFTWARE\eMicrosoft\*(Aq
.Ve
.PP
You can also display single values from within registry keys, for example:
.PP
.Vb 3
\& $ cvkey=\*(AqHKLM\eSOFTWARE\eMicrosoft\eWindows NT\eCurrentVersion\*(Aq
\& $ virt\-win\-reg Windows7 $cvkey ProductName
\& Windows 7 Enterprise
.Ve
.PP
\&\fI\-\-merge\fR を用いると、テキストの regedit ファイルを Windows レジストリーの中に結合できます:
.PP
.Vb 1
\& $ virt\-win\-reg \-\-merge Windows7 changes.reg
.Ve
.SS "注記"
.IX Subsection "注記"
This program is only meant for simple access to the registry.  If you want
to do complicated things with the registry, we suggest you download the
Registry hive files from the guest using \fBlibguestfs\fR\|(3) or \fBguestfish\fR\|(1)
and access them locally, eg. using \fBhivex\fR\|(3), \fBhivexsh\fR\|(1) or
\&\fBhivexregedit\fR\|(1).
.SH "オプション"
.IX Header "オプション"
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
簡単なヘルプを表示します。
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
バージョン番号を表示して、終了します。
.IP "\fB\-\-debug\fR" 4
.IX Item "--debug"
デバッグメッセージを有効にします。
.IP "\fB\-c \s-1URI\s0\fR" 4
.IX Item "-c URI"
.PD 0
.IP "\fB\-\-connect \s-1URI\s0\fR" 4
.IX Item "--connect URI"
.PD
libvirt を使用していると、指定された \fI\s-1URI\s0\fR に接続します。  省略すると、デフォルトの libvirt ハイパーバイザーに接続します。
.Sp
仮想マシンのブロックデバイスを直接指定すると、libvirt はまったく使用されません。
.IP "\fB\-\-format\fR raw" 4
.IX Item "--format raw"
Specify the format of disk images given on the command line.  If this is
omitted then the format is autodetected from the content of the disk image.
.Sp
If disk images are requested from libvirt, then this program asks libvirt
for this information.  In this case, the value of the format parameter is
ignored.
.Sp
If working with untrusted raw-format guest disk images, you should ensure
the format is always specified.
.IP "\fB\-\-merge\fR" 4
.IX Item "--merge"
In merge mode, this merges a textual regedit file into the Windows Registry
of the virtual machine.  If this flag is \fInot\fR given then virt-win-reg
displays or exports Registry entries instead.
.Sp
Note that \fI\-\-merge\fR is \fIunsafe\fR to use on live virtual machines, and will
result in disk corruption.  However exporting (without this flag)  is always
safe.
.IP "\fB\-\-encoding\fR UTF\-16LE|ASCII" 4
.IX Item "--encoding UTF-16LE|ASCII"
When merging (only), you may need to specify the encoding for strings to be
used in the hive file.  This is explained in detail in
\&\*(L"\s-1ENCODING STRINGS\*(R"\s0 in \fBWin::Hivex::Regedit\fR\|(3).
.Sp
The default is to use \s-1UTF\-16LE,\s0 which should work with recent versions of
Windows.
.IP "\fB\-\-unsafe\-printable\-strings\fR" 4
.IX Item "--unsafe-printable-strings"
When exporting (only), assume strings are \s-1UTF\-16LE\s0 and print them as strings
instead of hex sequences.  Remove the final zero codepoint from strings if
present.
.Sp
This is unsafe and does not preserve the fidelity of strings in the original
Registry for various reasons:
.RS 4
.IP "\(bu" 4
Assumes the original encoding is \s-1UTF\-16LE.\s0  \s-1ASCII\s0 strings and strings in
other encodings will be corrupted by this transformation.
.IP "\(bu" 4
Assumes that everything which has type 1 or 2 is really a string and that
everything else is not a string, but the type field in real Registries is
not reliable.
.IP "\(bu" 4
Loses information about whether a zero codepoint followed the string in the
Registry or not.
.RE
.RS 4
.Sp
This all happens because the Registry itself contains no information about
how strings are encoded (see \*(L"\s-1ENCODING STRINGS\*(R"\s0 in \fBWin::Hivex::Regedit\fR\|(3)).
.Sp
You should only use this option for quick hacking and debugging of the
Registry contents, and \fInever\fR use it if the output is going to be passed
into another program or stored in another Registry.
.RE
.SH "サポートされるシステム"
.IX Header "サポートされるシステム"
The program currently supports Windows NT-derived guests starting with
Windows \s-1XP\s0 through to at least Windows 8.
.PP
以下のレジストリキーがサポートされます:
.ie n .IP """HKEY_LOCAL_MACHINE\eSAM""" 4
.el .IP "\f(CWHKEY_LOCAL_MACHINE\eSAM\fR" 4
.IX Item "HKEY_LOCAL_MACHINESAM"
.PD 0
.ie n .IP """HKEY_LOCAL_MACHINE\eSECURITY""" 4
.el .IP "\f(CWHKEY_LOCAL_MACHINE\eSECURITY\fR" 4
.IX Item "HKEY_LOCAL_MACHINESECURITY"
.ie n .IP """HKEY_LOCAL_MACHINE\eSOFTWARE""" 4
.el .IP "\f(CWHKEY_LOCAL_MACHINE\eSOFTWARE\fR" 4
.IX Item "HKEY_LOCAL_MACHINESOFTWARE"
.ie n .IP """HKEY_LOCAL_MACHINE\eSYSTEM""" 4
.el .IP "\f(CWHKEY_LOCAL_MACHINE\eSYSTEM\fR" 4
.IX Item "HKEY_LOCAL_MACHINESYSTEM"
.ie n .IP """HKEY_USERS\e.DEFAULT""" 4
.el .IP "\f(CWHKEY_USERS\e.DEFAULT\fR" 4
.IX Item "HKEY_USERS.DEFAULT"
.ie n .IP """HKEY_USERS\e\fISID\fP""" 4
.el .IP "\f(CWHKEY_USERS\e\f(CISID\f(CW\fR" 4
.IX Item "HKEY_USERSSID"
.PD
ここで \fI\s-1SID\s0\fR は Windows User \s-1SID\s0 です (例: \f(CW\*(C`S\-1\-5\-18\*(C'\fR)。
.ie n .IP """HKEY_USERS\e\fIusername\fP""" 4
.el .IP "\f(CWHKEY_USERS\e\f(CIusername\f(CW\fR" 4
.IX Item "HKEY_USERSusername"
ここで \fIusername\fR はローカルユーザー名です (これは libguestfs の拡張です)。
.PP
You can use \f(CW\*(C`HKLM\*(C'\fR as a shorthand for \f(CW\*(C`HKEY_LOCAL_MACHINE\*(C'\fR, and \f(CW\*(C`HKU\*(C'\fR for
\&\f(CW\*(C`HKEY_USERS\*(C'\fR.
.PP
The literal keys \f(CW\*(C`HKEY_USERS\e$SID\*(C'\fR and \f(CW\*(C`HKEY_CURRENT_USER\*(C'\fR are not
supported (there is no \*(L"current user\*(R").
.SS "\s-1WINDOWS 8\s0"
.IX Subsection "WINDOWS 8"
Windows 8 \*(L"fast startup\*(R" can prevent virt-win-reg from being able to edit
the Registry.  See \*(L"\s-1WINDOWS HIBERNATION AND WINDOWS 8 FAST
STARTUP\*(R"\s0 in \fBguestfs\fR\|(3).
.SH "エンコーディング"
.IX Header "エンコーディング"
\&\f(CW\*(C`virt\-win\-reg\*(C'\fR expects that regedit files have already been reencoded in
the local encoding.  Usually on Linux hosts, this means \s-1UTF\-8\s0 with
Unix-style line endings.  Since Windows regedit files are often in \s-1UTF\-16LE\s0
with Windows-style line endings, you may need to reencode the whole file
before or after processing.
.PP
To reencode a file from Windows format to Linux (before processing it with
the \fI\-\-merge\fR option), you would do something like this:
.PP
.Vb 1
\& iconv \-f utf\-16le \-t utf\-8 < win.reg | dos2unix > linux.reg
.Ve
.PP
To go in the opposite direction, after exporting and before sending the file
to a Windows user, do something like this:
.PP
.Vb 1
\& unix2dos linux.reg | iconv \-f utf\-8 \-t utf\-16le > win.reg
.Ve
.PP
エンコーディングの詳細は \fBWin::Hivex::Regedit\fR\|(3) を参照してください。
.PP
If you are unsure about the current encoding, use the \fBfile\fR\|(1) command.
Recent versions of Windows regedit.exe produce a \s-1UTF\-16LE\s0 file with
Windows-style (\s-1CRLF\s0) line endings, like this:
.PP
.Vb 3
\& $ file software.reg
\& software.reg: Little\-endian UTF\-16 Unicode text, with very long lines,
\& with CRLF line terminators
.Ve
.PP
This file would need conversion before you could \fI\-\-merge\fR it.
.SH "CurrentControlSet など。"
.IX Header "CurrentControlSet など。"
Registry keys like \f(CW\*(C`CurrentControlSet\*(C'\fR don’t really exist in the Windows
Registry at the level of the hive file, and therefore you cannot modify
these.
.PP
\&\f(CW\*(C`CurrentControlSet\*(C'\fR is usually an alias for \f(CW\*(C`ControlSet001\*(C'\fR.  In some
circumstances it might refer to another control set.  The way to find out is
to look at the \f(CW\*(C`HKLM\eSYSTEM\eSelect\*(C'\fR key:
.PP
.Vb 6
\& # virt\-win\-reg WindowsGuest \*(AqHKLM\eSYSTEM\eSelect\*(Aq
\& [HKEY_LOCAL_MACHINE\eSYSTEM\eSelect]
\& "Current"=dword:00000001
\& "Default"=dword:00000001
\& "Failed"=dword:00000000
\& "LastKnownGood"=dword:00000002
.Ve
.PP
\&\*(L"Current\*(R" is the one which Windows will choose when it boots.
.PP
Similarly, other \f(CW\*(C`Current...\*(C'\fR keys in the path may need to be replaced.
.SH "レジストリーキーおよび値の削除方法"
.IX Header "レジストリーキーおよび値の削除方法"
レジストリキー全体を削除するには、この構文を使用します:
.PP
.Vb 1
\& [\-HKEY_LOCAL_MACHINE\eFoo]
.Ve
.PP
キーにある単一の値を削除するには、この構文を使用します:
.PP
.Vb 2
\& [HKEY_LOCAL_MACHINE\eFoo]
\& "Value"=\-
.Ve
.SH "Windows のヒント"
.IX Header "Windows のヒント"
Note that some of these tips modify the guest disk image.  The guest \fImust\fR
be shut off, else you will get disk corruption.
.SS "\s-1RUNNING A BATCH SCRIPT WHEN A USER LOGS IN\s0"
.IX Subsection "RUNNING A BATCH SCRIPT WHEN A USER LOGS IN"
Prepare a \s-1DOS\s0 batch script, VBScript or executable.  Upload this using
\&\fBguestfish\fR\|(1).  For this example the script is called \f(CW\*(C`test.bat\*(C'\fR and it
is uploaded into \f(CW\*(C`C:\e\*(C'\fR:
.PP
.Vb 1
\& guestfish \-i \-d WindowsGuest upload test.bat /test.bat
.Ve
.PP

ChangeLog
commit 1f026244ba5891a00617f4ab8c41b8680ff8c497
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Thu Jan 17 12:06:36 2019 +0000

    Version 1.40.1.

commit ae2aa14527605ffb89d86bf93bf8ca4f969118e1
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Thu Jan 17 12:03:07 2019 +0000

    valgrind: Add workaround for slow loading of debuginfo by valgrind.
    
    See nbdkit commit
    https://github.com/libguestfs/nbdkit/commit/bd32d899edb4c592a2df0ce4b26a2cf8396eba1d
    and https://bugzilla.redhat.com/show_bug.cgi?id=1662656
    
    (cherry picked from commit ea7c13fed7922b233db8382b1348dfb436e3ba85)

commit 1de72f8f00600bcaa76742dcfd366fb702eef3c2
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Tue Jan 15 14:39:41 2019 +0000

    Version 1.40.0.

commit 6151d8694d9db0d709a9ad807a5ca6a098d6502c
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Wed Jan 16 11:49:45 2019 +0000

    po: Drop ca and pt_BR from linguas.
    
    Commit 4252a490bb117940c1d38c838345ed110457386e dropped the PO files
    for ca (Catalan) and pt_BR (Brazilian Portuguese).  However it did not
    drop these languages from the linguas line in the Makefile resulting
    in a build failure.

commit d1ff62f3e08fd8af5e01cff8045f0a01a5aa0597
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Wed Jan 16 11:39:33 2019 +0000

    configure: Move remaining detection of p2v features to m4/guestfs-v2v.m4.
    
    ./configure output now contains a distinct section for v2v and p2v:
    
      --- Checking the virt-v2v and virt-p2v dependencies ---
      checking for the nbdkit python plugin name... python3
      checking for --with-gtk option... 2
      checking for GTK... yes
      checking for DBUS... yes
      checking if we can build virt-p2v... yes, with Gtk 2
    
    Thanks: Pino Toscano.

commit 61b86bac2e2071c1760b46d13195b4b8e24abd7d
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Tue Jan 8 12:11:20 2019 +0000

    v2v: -o rhv-upload: Allow configure to set the nbdkit Python version.
    
    No functional change, but it does allow downstream distributions to
    adjust the nbdkit Python plugin used by virt-v2v -o rhv-upload mode:
    
    ./configure --with-virt-v2v-nbdkit-python-plugin=...

commit 090f5589e7b838375bc5333fe0f7972630001348
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Tue Jan 15 14:43:28 2019 +0000

    po-docs: uk: Fix inconsistent use of \n in translation.
    
    Invalid po file ./uk.po:
    ./uk.po:90428: 'msgid' and 'msgstr' entries do not both begin with '\n'
    msgfmt: found 1 fatal error

commit 1db01404281b2db740f8b854751efc083dcbffa1
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Tue Jan 15 13:43:29 2019 +0000

    docs: Finalize release notes for 1.40.
    
    Updates commit 0ce8ecdf6627bedfebc235fc4fd69f435a480e88.

commit b8882fed975a09bf7aa14f09eb343ac74c899e71
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Tue Jan 15 13:42:18 2019 +0000

    website: Update website for 1.40.

commit cf6b527824b2a8dc6e8bc65e38ebdceb227e6db1
Author: Pino Toscano <ptoscano@redhat.com>
Date:   Mon Jan 14 17:07:45 2019 +0100

    inspect: fix inspection of partition-less devices (RHBZ#1661038)
    
    When parsing "xdev"-kind devices, do not assume that the partition
    number can be converted to integer: re_xdev accepts an empty part of the
    partition number, so just handle as it is, as string.
    
    This fixes a regression due to the conversion of the inspection code to
    OCaml, as the old C version did not have this issue.

commit d3ff88c894b0ddadea005b8563b5793fa6b5b9be
Author: Pino Toscano <ptoscano@redhat.com>
Date:   Fri Jan 11 17:00:11 2019 +0100

    OCaml: use the new behaviour of Std_utils.which
    
    Since Std_utils.which can handle relative/absolute paths, remove the
    manual checks, and use Std_utils.which directly.

commit ebe2c7b45832f41e75deb73d784f05ebc87c9b71
Author: Pino Toscano <ptoscano@redhat.com>
Date:   Fri Jan 11 15:59:03 2019 +0100

    mlstdutils: allow relative/absolute paths for Std_utils.which
    
    Make Std_utils.which behave a bit more like which(1), checking the
    existance of relative/absolute paths specified.

commit 88a754bd072d26b660964e9883303a2939d28fc5
Author: Pino Toscano <ptoscano@redhat.com>
Date:   Fri Jan 11 15:19:48 2019 +0100

    mlstdutils: add a very simple test for Std_utils.which

commit 27cd2888c31dc3b93a1c8973010509d69ac38778
Author: Pino Toscano <ptoscano@redhat.com>
Date:   Mon Jan 14 15:18:34 2019 +0100

    automake2junit.ml: string -> bytes
    
    Make it usable OOTB with OCaml >= 4.02.0.

commit 7f97433a3c067ea7c2a2a159b09720c0772605c8
Author: Yuri Chornoivan <yurchor@ukr.net>
Date:   Fri Jan 11 17:14:21 2019 +0000

    Fix miscellaneous typos in comments.

commit 4040c7de93053f93674dac3fc8523feef207fa5f
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Tue Jan 8 18:29:14 2019 +0000

    Update gnulib to latest version.

commit 4252a490bb117940c1d38c838345ed110457386e
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Tue Jan 8 16:11:31 2019 +0000

    Update PO files from Zanata.

commit fc028bf57a3ff128d21b904583f9ea02f672ed5b
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Tue Jan 8 14:03:25 2019 +0000

    v2v: -o openstack: Don't echo full commands (RHBZ#1664310).
    
    They can contain passwords or tokens if for example the
    ‘-oo os-password’ option is used.
    
    Thanks: Tomáš Golembiovský, Brett Thurber.

commit 05d4fcb64d98d4ff1d57560c566ca8e66b695277
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Tue Jan 8 11:58:30 2019 +0000

    Update copyright dates for 2019.
    
    This command run over the source:
    
    perl -pi.bak -e 's/(20[01][0-9])-2018/$1-2019/g' `git ls-files`

commit 0ce8ecdf6627bedfebc235fc4fd69f435a480e88
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Mon Jan 7 16:48:15 2019 +0000

    Update release notes for 1.40 release.

commit 492a945791b43f80a769a53e60d0899b3d7c60ab

po-docs/uk/virt-win-reg.pl
=encoding utf8

=head1 НАЗВА

virt-win-reg - Export and merge Windows Registry entries from a Windows
guest

=head1 КОРОТКИЙ ОПИС

 virt-win-reg domname 'HKLM\Path\To\Subkey'

 virt-win-reg domname 'HKLM\Path\To\Subkey' name

 virt-win-reg domname 'HKLM\Path\To\Subkey' @

 virt-win-reg --merge domname [input.reg ...]

 virt-win-reg [--options] disk.img ... # instead of domname

=head1 ПОПЕРЕДЖЕННЯ

You must I<not> use C<virt-win-reg> with the I<--merge> option on live
virtual machines.  If you do this, you I<will> get irreversible disk
corruption in the VM.  C<virt-win-reg> tries to stop you from doing this,
but doesn't catch all cases.

Modifying the Windows Registry is an inherently risky operation.  The format
is deliberately obscure and undocumented, and Registry changes can leave the
system unbootable.  Therefore when using the I<--merge> option, make sure
you have a reliable backup first.

=head1 ОПИС

This program can export and merge Windows Registry entries from a Windows
guest.

The first parameter is the libvirt guest name or the raw disk image of a
Windows guest.

If I<--merge> is I<not> specified, then the chosen registry key is
displayed/exported (recursively).  For example:

 $ virt-win-reg Windows7 'HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft'

You can also display single values from within registry keys, for example:

 $ cvkey='HKLM\SOFTWARE\Microsoft\Windows NT\CurrentVersion'
 $ virt-win-reg Windows7 $cvkey ProductName
 Windows 7 Enterprise

With I<--merge>, you can merge a textual regedit file into the Windows
Registry:

 $ virt-win-reg --merge Windows7 changes.reg

=head2 ПРИМІТКА

This program is only meant for simple access to the registry.  If you want
to do complicated things with the registry, we suggest you download the
Registry hive files from the guest using L<libguestfs(3)> or L<guestfish(1)>
and access them locally, eg. using L<hivex(3)>, L<hivexsh(1)> or
L<hivexregedit(1)>.

=head1 ПАРАМЕТРИ

=over 4

=item B<--help>

Показати коротку довідку.

=item B<--version>

Показати дані щодо версії і завершити роботу.

=item B<--debug>

Увімкнути показ діагностичних повідомлень.

=item B<-c адреса>

=item B<--connect адреса>

Якщо використовується libvirt, встановити з’єднання з вказаним I<URI>. Якщо
пропущено, з’єднання буде встановлено з типовим гіпервізором libvirt.

If you specify guest block devices directly, then libvirt is not used at
all.

=item B<--format> raw

Specify the format of disk images given on the command line.  If this is
omitted then the format is autodetected from the content of the disk image.

If disk images are requested from libvirt, then this program asks libvirt
for this information.  In this case, the value of the format parameter is
ignored.

If working with untrusted raw-format guest disk images, you should ensure
the format is always specified.

=item B<--merge>

In merge mode, this merges a textual regedit file into the Windows Registry
of the virtual machine.  If this flag is I<not> given then virt-win-reg
displays or exports Registry entries instead.

Note that I<--merge> is I<unsafe> to use on live virtual machines, and will
result in disk corruption.  However exporting (without this flag)  is always
safe.

=item B<--encoding> UTF-16LE|ASCII

When merging (only), you may need to specify the encoding for strings to be
used in the hive file.  This is explained in detail in
L<Win::Hivex::Regedit(3)/ENCODING STRINGS>.

The default is to use UTF-16LE, which should work with recent versions of
Windows.

=item B<--unsafe-printable-strings>

When exporting (only), assume strings are UTF-16LE and print them as strings
instead of hex sequences.  Remove the final zero codepoint from strings if
present.

This is unsafe and does not preserve the fidelity of strings in the original
Registry for various reasons:

=over 4

=item *

Assumes the original encoding is UTF-16LE.  ASCII strings and strings in
other encodings will be corrupted by this transformation.

=item *

Assumes that everything which has type 1 or 2 is really a string and that
everything else is not a string, but the type field in real Registries is
not reliable.

=item *

Loses information about whether a zero codepoint followed the string in the
Registry or not.

=back

This all happens because the Registry itself contains no information about
how strings are encoded (see L<Win::Hivex::Regedit(3)/ENCODING STRINGS>).

You should only use this option for quick hacking and debugging of the
Registry contents, and I<never> use it if the output is going to be passed
into another program or stored in another Registry.

=back

=head1 ПІДТРИМУВАНІ СИСТЕМИ

The program currently supports Windows NT-derived guests starting with
Windows XP through to at least Windows 8.

The following Registry keys are supported:

=over 4

=item C<HKEY_LOCAL_MACHINE\SAM>

=item C<HKEY_LOCAL_MACHINE\SECURITY>

=item C<HKEY_LOCAL_MACHINE\SOFTWARE>

=item C<HKEY_LOCAL_MACHINE\SYSTEM>

=item C<HKEY_USERS\.DEFAULT>

=item C<HKEY_USERS\I<SID>>

where I<SID> is a Windows User SID (eg. C<S-1-5-18>).

=item C<HKEY_USERS\I<користувач>>

where I<username> is a local user name (this is a libguestfs extension).

=back

You can use C<HKLM> as a shorthand for C<HKEY_LOCAL_MACHINE>, and C<HKU> for
C<HKEY_USERS>.

The literal keys C<HKEY_USERS\$SID> and C<HKEY_CURRENT_USER> are not
supported (there is no "current user").

=head2 WINDOWS 8

Windows 8 "fast startup" can prevent virt-win-reg from being able to edit
the Registry.  See L<guestfs(3)/WINDOWS HIBERNATION AND WINDOWS 8 FAST
STARTUP>.

=head1 КОДУВАННЯ

C<virt-win-reg> expects that regedit files have already been reencoded in
the local encoding.  Usually on Linux hosts, this means UTF-8 with
Unix-style line endings.  Since Windows regedit files are often in UTF-16LE
with Windows-style line endings, you may need to reencode the whole file
before or after processing.

To reencode a file from Windows format to Linux (before processing it with
the I<--merge> option), you would do something like this:

 iconv -f utf-16le -t utf-8 < win.reg | dos2unix > linux.reg

To go in the opposite direction, after exporting and before sending the file
to a Windows user, do something like this:

 unix2dos linux.reg | iconv -f utf-8 -t utf-16le > win.reg

For more information about encoding, see L<Win::Hivex::Regedit(3)>.

If you are unsure about the current encoding, use the L<file(1)> command.
Recent versions of Windows regedit.exe produce a UTF-16LE file with
Windows-style (CRLF) line endings, like this:

 $ file software.reg
 software.reg: Little-endian UTF-16 Unicode text, with very long lines,
 with CRLF line terminators

This file would need conversion before you could I<--merge> it.

=head1 CurrentControlSet etc.

Registry keys like C<CurrentControlSet> don't really exist in the Windows
Registry at the level of the hive file, and therefore you cannot modify
these.

C<CurrentControlSet> is usually an alias for C<ControlSet001>.  In some
circumstances it might refer to another control set.  The way to find out is
to look at the C<HKLM\SYSTEM\Select> key:

 # virt-win-reg WindowsGuest 'HKLM\SYSTEM\Select'
 [HKEY_LOCAL_MACHINE\SYSTEM\Select]
 "Current"=dword:00000001
 "Default"=dword:00000001
 "Failed"=dword:00000000
 "LastKnownGood"=dword:00000002

"Current" is the one which Windows will choose when it boots.

Similarly, other C<Current...> keys in the path may need to be replaced.

=head1 DELETING REGISTRY KEYS AND VALUES

To delete a whole registry key, use the syntax:

 [-HKEY_LOCAL_MACHINE\Foo]

To delete a single value within a key, use the syntax:

 [HKEY_LOCAL_MACHINE\Foo]
 "Value"=-

=head1 WINDOWS TIPS

Note that some of these tips modify the guest disk image.  The guest I<must>
be shut off, else you will get disk corruption.

=head2 RUNNING A BATCH SCRIPT WHEN A USER LOGS IN

Prepare a DOS batch script, VBScript or executable.  Upload this using
L<guestfish(1)>.  For this example the script is called C<test.bat> and it
is uploaded into C<C:\>:

 guestfish -i -d WindowsGuest upload test.bat /test.bat

Prepare a regedit file containing the registry change:

 cat > test.reg <<'EOF'
 [HKLM\Software\Microsoft\Windows\CurrentVersion\RunOnce]
 "Test"="c:\\test.bat"
 EOF

In this example we use the key C<RunOnce> which means that the script will
run precisely once when the first user logs in.  If you want it to run every
time a user logs in, replace C<RunOnce> with C<Run>.

Оновіть регістр:

 virt-win-reg --merge WindowsGuest test.reg

=head2 INSTALLING A SERVICE

This section assumes you are familiar with Windows services, and you either
have a program which handles the Windows Service Control Protocol directly
or you want to run any program using a service wrapper like SrvAny or the
free RHSrvAny.

First upload the program and optionally the service wrapper.  In this case
the test program is called C<test.exe> and we are using the RHSrvAny
wrapper:

 guestfish -i -d WindowsGuest <<EOF
   upload rhsrvany.exe /rhsrvany.exe
   upload test.exe /test.exe
 EOF

Prepare a regedit file containing the registry changes.  In this example,
the first registry change is needed for the service itself or the service
wrapper (if used).  The second registry change is only needed because I am
using the RHSrvAny service wrapper.

 cat > service.reg <<'EOF'
 [HKLM\SYSTEM\ControlSet001\services\RHSrvAny]
 "Type"=dword:00000010
 "Start"=dword:00000002
 "ErrorControl"=dword:00000001
 "ImagePath"="c:\\rhsrvany.exe"
 "DisplayName"="RHSrvAny"
 "ObjectName"="NetworkService"
 
 [HKLM\SYSTEM\ControlSet001\services\RHSrvAny\Parameters]
 "CommandLine"="c:\\test.exe"
 "PWD"="c:\\Temp"
 EOF

Нотатки:

=over 4

=item *

For use of C<ControlSet001> see the section above in this manual page.  You
may need to adjust this according to the control set that is in use by the
guest.

=item *

C<"ObjectName"> controls the privileges that the service will have.  An
alternative is C<"ObjectName"="LocalSystem"> which would be the most
privileged account.

=item *

For the meaning of the magic numbers, see this Microsoft KB article:
L<http://support.microsoft.com/kb/103000>.

=back

Оновіть регістр:

 virt-win-reg --merge WindowsGuest service.reg

=head1 ЕКРАНУВАННЯ СИМВОЛІВ У ОБОЛОНЦІ

Be careful when passing parameters containing C<\> (backslash) in the
shell.  Usually you will have to use 'single quotes' or double backslashes
(but not both) to protect them from the shell.

Paths and value names are case-insensitive.

Libvirt guest names can contain arbitrary characters, some of which have
meaning to the shell such as C<#> and space.  You may need to quote or
escape these characters on the command line.  See the shell manual page
L<sh(1)> for details.

=head1 ТАКОЖ ПЕРЕГЛЯНЬТЕ

L<hivex(3)>, L<hivexsh(1)>, L<hivexregedit(1)>, L<guestfs(3)>,
L<guestfish(1)>, L<virt-cat(1)>, L<Sys::Guestfs(3)>, L<Win::Hivex(3)>,
L<Win::Hivex::Regedit(3)>, L<Sys::Virt(3)>, L<http://libguestfs.org/>.

=head1 АВТОР

Richard W.M. Jones L<http://people.redhat.com/~rjones/>

=head1 АВТОРСЬКІ ПРАВА

©Red Hat Inc., 2010

